### DISCUSSION TITLE
How to enable HTTPS is AdonisJS 5 ?

### DISCUSSION THREAD

**Discussion by ghost** [adonisjs/core#1742] [Help] [answered]
I was using this code in express to enable https:

```
const fs = require('fs');
const path = require('path');
const https = require('https');

const express = require('express');
const app = express();

const privateKey = fs.readFileSync('sslCert/server.key', 'utf8');
const certificate = fs.readFileSync('sslCert/server.crt', 'utf8');
const credentials = {key: privateKey, cert: certificate};

const httpsServer = https.createServer(credentials, app);

httpsServer.listen(8443); // # https://localhost:8443/
console.log('Listening for https on port 8443...');

```

I have the certificate and the private key stored on the server in a separate file , i made the cert and the key using this command:
```
openssl req -x509 -sha256 -nodes -days 99999 -newkey rsa:4096 -keyout
private.key -out certificate.crt
```

Now how do i use them to enable ssl ? thanks for help and support

**Comment by thetutlage** [comment, marked as answer]
Inside `server.ts`, you can pass a callback to the `start` method to return a custom Node.js server instance. Checkout the following example.


```ts
import { createServer } from "https";
import { Ignitor } from "@adonisjs/core/build/src/Ignitor";

new Ignitor(__dirname).httpServer().start((handle) => {
  return createServer(
    {
      key: "",
      cert: "",
    },
    handle
  );
});
```

**Reply by ghost** [reply]
Thanks for help , my app name is "AdonisJS" , i have the server file "AdonisJS/server.ts" , and i have the "AdonisJS/sslCert" folder that contain "server.key" and "server.crt" , did this:

```
import 'reflect-metadata'
import sourceMapSupport from 'source-map-support';
import { join } from 'path';
import { readFileSync } from 'fs';
import { createServer } from 'https';
import { Ignitor } from '@adonisjs/core/build/src/Ignitor';

sourceMapSupport.install({ handleUncaughtExceptions: false })

const privateKey = readFileSync(join(__dirname + '/sslCert/server.key'), 'utf8');
const certificate = readFileSync(join(__dirname + '/sslCert/server.crt'), 'utf8');
const credentials = {key: privateKey, cert: certificate};

new Ignitor(__dirname)
  .httpServer()
  .start((handle) => {
    return createServer(credentials, handle);
  })
  .catch(console.error)


```

but in the terminal after running the app "node ace serve --dev" it keeps saying:

```
ℹ  info      cleaning up build directory build
ℹ  info      copy .env,resources/views/**/*.edge,public/**,ace build
☐  pending   compiling typescript source files
✔  success   built successfully
ℹ  info      copy .adonisrc.json build
✔  create    ace-manifest.json
ℹ  info      starting http server
internal/fs/utils.js:299
    throw err;
    ^

Error: ENOENT: no such file or directory, open '/home/rootx/Documents/nodeJS/AdonisJS/build/sslCert/server.key'
    at Object.openSync (fs.js:474:3)
    at Object.readFileSync (fs.js:375:35)
    at Object.<anonymous> (/home/rootx/Documents/nodeJS/AdonisJS/server.ts:22:20)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/home/rootx/Documents/nodeJS/AdonisJS/build/sslCert/server.key'
}

```

so i replaced the sslCert folder inside "AdonisJS/build" folder , but after i rerun the server the sslCert disappears , so what should i do , where should i place the sslCert folder ?

**Reply by thetutlage** [reply]
Write the following like of code inside `metaFile` array in `.adonisrc.json` file.

```json
"metaFiles": [
  "sslCert/*"
]
```

This will copy your `sslCert` folder to the `build` when starting the app and the server will be able to find it

**Reply by ghost** [reply]
Thank you so much , it worked.

**Reply by aslampatel** [reply]
i used this code in adonis js 4.1  for http to https but not working can  what's wrong on this code
in hook file 
const { hooks }      = require('@adonisjs/ignitor')
const Helpers        = use('Helpers')
const Server         = use('Server')
const Env            = use('Env')
const https       = require('https')
const fs          = require('fs')
hooks.before.httpServer(() => {
  
        var certOptions = {
          key: fs.readFileSync('./ssl/serverkey.key'),
          cert: fs.readFileSync('./ssl/servercrt.crt'),
        }
        return https.createServer(certOptions)
})

my server.js file code is i am calling hook in preload but ssl not working

new Ignitor(require('@adonisjs/fold'))
.appRoot(__dirname)
.preLoad('start/hook')
.fireHttpServer()
.catch(console.error)

**Reply by ghiovani666** [reply]
used this code in adonis js 4.1 for http to https



const { Ignitor } = require('@adonisjs/ignitor')

// ::=======================================================================
const https = require('https')
const fs = require('fs')

var certOptions = {
  key: fs.readFileSync('./sslCert/server.key'),
  cert: fs.readFileSync('./sslCert/server.cer'),
}

// ::=======================================================================
new Ignitor(require('@adonisjs/fold'))
  .appRoot(__dirname)
  .fireHttpServer((handle) => {
    return https.createServer(certOptions, handle);
  })
  .catch(console.error)

**Comment by guoyunhe** [comment]
In my case, I want to use `.env` file to store the location of certificates. However, the following code:

```ts
// server.ts
import { Ignitor } from '@adonisjs/core/build/standalone';
import Env from '@ioc:Adonis/Core/Env';
import { readFileSync } from 'fs';
import { createServer as createHttpServer } from 'http';
import { createServer as createHttpsServer } from 'https';
import 'reflect-metadata';
import sourceMapSupport from 'source-map-support';

sourceMapSupport.install({ handleUncaughtExceptions: false });

new Ignitor(__dirname).httpServer().start((handle) => {
  if (Env.get('SSL_KEY')) {
    return createHttpsServer(
      {
        key: readFileSync(Env.get('SSL_KEY')),
        cert: readFileSync(Env.get('SSL_CERT')),
      },
      handle
    );
  } else {
    return createHttpServer(handle);
  }
});
```

Throw errors:

```
import Env from '@ioc:Adonis/Core/Env';
^
TypeError: global[Symbol.for(...)] is not a function
    at Object.<anonymous> (/home/guo/git/rgi-api/server.ts:14:1)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at Module._compile (/home/guo/git/rgi-api/node_modules/.pnpm/pirates@4.0.5/node_modules/pirates/lib/index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Object.newLoader [as .ts] (/home/guo/git/rgi-api/node_modules/.pnpm/pirates@4.0.5/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Function.Module._load (node:internal/modules/cjs/loader:922:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
```

**Reply by mikesaintsg** [reply]
You're trying to access the ioc container with a top level import before it has been registered

Remove the top level import of Env and before the if statement add:
```js
const { default: Env } = await import('@ioc:Adonis/Core/Env')
```
